File: frontend/src/types/risk.ts
/**
 * Risk Management TypeScript Types
 * Complete type definitions for risk metrics, violations, limits, and emergency stops.
 */

// ===========================
// ENUM TYPES
// ===========================

export enum RiskStatus {
  NORMAL = 'normal',
  WARNING = 'warning',
  CRITICAL = 'critical',
  BREACHED = 'breached',
}

export enum ViolationType {
  WARNING = 'warning',
  BREACH = 'breach',
}

export enum Severity {
  LOW = 'low',
  MEDIUM = 'medium',
  HIGH = 'high',
  CRITICAL = 'critical',
}

export enum EmergencyStopStatus {
  ACTIVE = 'active',
  RESOLVED = 'resolved',
}

// ===========================
// CORE DATA TYPES
// ===========================

export interface RiskMetric {
  id: string;
  user_id: string;
  metric_name: string;
  current_value: number;
  limit_value: number;
  percent_used: number;
  status: RiskStatus;
  last_updated: string;
  created_at: string;
}

export interface RiskViolation {
  id: string;
  user_id: string;
  metric_name: string;
  violation_type: ViolationType;
  current_value: number;
  limit_value: number;
  severity: Severity;
  message: string;
  resolved: boolean;
  resolved_at?: string;
  created_at: string;
}

export interface RiskLimit {
  id: string;
  user_id: string;
  limit_name: string;
  limit_value: number;
  warning_threshold: number;
  critical_threshold: number;
  enabled: boolean;
  created_at: string;
  updated_at: string;
  updated_by?: string;
}

export interface EmergencyStop {
  id: string;
  user_id: string;
  triggered_by: string;
  reason: string;
  strategies_stopped: number;
  orders_cancelled: number;
  status: EmergencyStopStatus;
  triggered_at: string;
  resolved_at?: string;
  resolved_by?: string;
}

// ===========================
// REQUEST TYPES
// ===========================

export interface UpdateRiskLimitRequest {
  limit_name: string;
  limit_value: number;
  warning_threshold?: number;
  critical_threshold?: number;
  enabled?: boolean;
}

export interface TriggerEmergencyStopRequest {
  reason: string;
  stop_all_strategies?: boolean;
  cancel_all_orders?: boolean;
  notify_admin?: boolean;
}

// ===========================
// DASHBOARD DATA TYPES
// ===========================

export interface RiskDashboardData {
  metrics: RiskMetric[];
  recent_violations: RiskViolation[];
  active_limits: RiskLimit[];
  emergency_status: EmergencyStop | null;
  summary: {
    total_metrics: number;
    breached_metrics: number;
    warning_metrics: number;
    critical_metrics: number;
    recent_violations_count: number;
    is_emergency_active: boolean;
  };
}

// ===========================
// WEBSOCKET EVENT TYPES
// ===========================

export interface RiskMetricUpdate {
  type: 'risk_metric_update';
  data: RiskMetric;
}

export interface RiskViolationAlert {
  type: 'risk_violation_alert';
  data: RiskViolation;
}

export interface EmergencyStopEvent {
  type: 'emergency_stop_event';
  data: EmergencyStop;
}

export type RiskWebSocketEvent = RiskMetricUpdate | RiskViolationAlert | EmergencyStopEvent;

// ===========================
// UI COMPONENT PROPS
// ===========================

export interface RiskDashboardProps {
  className?: string;
  showKillSwitch?: boolean;
  autoRefresh?: boolean;
  refreshInterval?: number;
}

export interface KillSwitchButtonProps {
  className?: string;
  size?: 'small' | 'medium' | 'large';
  disabled?: boolean;
  onConfirm?: () => void;
  onCancel?: () => void;
}

export interface RiskLimitsConfigProps {
  className?: string;
  editable?: boolean;
  onSave?: (limits: RiskLimit[]) => void;
}

export interface RiskMetricCardProps {
  metric: RiskMetric;
  className?: string;
  showDetails?: boolean;
}

export interface RiskViolationListProps {
  violations: RiskViolation[];
  className?: string;
  showResolved?: boolean;
  onResolve?: (violationId: string) => void;
}

// ===========================
// API ERROR TYPES
// ===========================

export interface RiskAPIError {
  detail: string;
  error_code?: string;
  timestamp: string;
}

// ===========================
// UTILITY TYPES
// ===========================

export interface RiskStatusConfig {
  status: RiskStatus;
  color: string;
  icon: string;
  description: string;
}

export interface SeverityConfig {
  severity: Severity;
  color: string;
  priority: number;
  description: string;
}

// ===========================
// FORM TYPES
// ===========================

export interface RiskLimitFormData {
  limit_name: string;
  limit_value: number;
  warning_threshold: number;
  critical_threshold: number;
  enabled: boolean;
}

export interface EmergencyStopFormData {
  reason: string;
  stop_all_strategies: boolean;
  cancel_all_orders: boolean;
  notify_admin: boolean;
}

// ===========================
// CONSTANTS
// ===========================

export const RISK_STATUS_CONFIGS: Record<RiskStatus, RiskStatusConfig> = {
  [RiskStatus.NORMAL]: {
    status: RiskStatus.NORMAL,
    color: '#52c41a',
    icon: 'CheckCircleOutlined',
    description: 'All metrics within normal limits',
  },
  [RiskStatus.WARNING]: {
    status: RiskStatus.WARNING,
    color: '#faad14',
    icon: 'WarningOutlined',
    description: 'Approaching risk limit',
  },
  [RiskStatus.CRITICAL]: {
    status: RiskStatus.CRITICAL,
    color: '#fa8c16',
    icon: 'ExclamationCircleOutlined',
    description: 'Very close to risk limit',
  },
  [RiskStatus.BREACHED]: {
    status: RiskStatus.BREACHED,
    color: '#ff4d4f',
    icon: 'CloseCircleOutlined',
    description: 'Risk limit exceeded',
  },
};

export const SEVERITY_CONFIGS: Record<Severity, SeverityConfig> = {
  [Severity.LOW]: {
    severity: Severity.LOW,
    color: '#52c41a',
    priority: 1,
    description: 'Minor violation, monitor closely',
  },
  [Severity.MEDIUM]: {
    severity: Severity.MEDIUM,
    color: '#faad14',
    priority: 2,
    description: 'Moderate violation, action recommended',
  },
  [Severity.HIGH]: {
    severity: Severity.HIGH,
    color: '#fa8c16',
    priority: 3,
    description: 'Serious violation, immediate attention required',
  },
  [Severity.CRITICAL]: {
    severity: Severity.CRITICAL,
    color: '#ff4d4f',
    priority: 4,
    description: 'Critical violation, emergency response needed',
  },
};

// ===========================
// DEFAULT VALUES
// ===========================

export const DEFAULT_RISK_LIMITS: Partial<RiskLimit> = {
  warning_threshold: 80,
  critical_threshold: 95,
  enabled: true,
};

export const DEFAULT_REFRESH_INTERVAL = 5000; // 5 seconds
export const MAX_RECENT_VIOLATIONS = 10;
export const EMERGENCY_CONFIRM_TIMEOUT = 10000; // 10 seconds

File: backend/models/risk.py
"""
Pydantic models for risk management.
Comprehensive type-safe models for risk metrics, violations, limits, and emergency stops.
"""

from datetime import datetime
from decimal import Decimal
from enum import Enum
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, field_serializer, field_validator


class RiskStatus(str, Enum):
    """Risk metric status levels"""

    NORMAL = "normal"
    WARNING = "warning"
    CRITICAL = "critical"
    BREACHED = "breached"


class ViolationType(str, Enum):
    """Type of risk violation"""

    WARNING = "warning"
    BREACH = "breach"


class Severity(str, Enum):
    """Violation severity levels"""

    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"
    CRITICAL = "critical"


class EmergencyStopStatus(str, Enum):
    """Emergency stop status"""

    ACTIVE = "active"
    RESOLVED = "resolved"


# ===========================
# REQUEST MODELS
# ===========================


class UpdateRiskLimitRequest(BaseModel):
    """Request to update a risk limit"""

    limit_value: Decimal = Field(..., gt=0, description="New limit value")
    warning_threshold: Decimal | None = Field(
        80.0, ge=0, le=100, description="Warning threshold %"
    )
    critical_threshold: Decimal | None = Field(
        95.0, ge=0, le=100, description="Critical threshold %"
    )
    enabled: bool | None = Field(True, description="Whether limit is enabled")

    @field_validator("critical_threshold")
    @classmethod
    def validate_thresholds(cls, v, info):
        if "warning_threshold" in info.data and v <= info.data["warning_threshold"]:
            raise ValueError("Critical threshold must be greater than warning threshold")
        return v


class TriggerEmergencyStopRequest(BaseModel):
    """Request to trigger emergency stop"""

    reason: str = Field(
        ..., min_length=10, max_length=500, description="Reason for emergency stop"
    )


# ===========================
# RESPONSE MODELS
# ===========================


# V11 prep / Wave-65 (VV-2 closure, 2026-05-03): emit Decimals as JSON
# numbers, not strings.  Pydantic v2's default Decimal -> JSON is a
# string, which broke the frontend's `RiskMetric.current_value: number`
# type contract.  Frontend was masking with `parseFloat(x.toString())`
# in 3 components.  field_serializer returns float on the wire while
# preserving Decimal precision in Python.
def _decimal_to_float(value: Decimal | None) -> float | None:
    return float(value) if value is not None else None


class RiskMetric(BaseModel):
    """Real-time risk metric"""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    user_id: int  # Changed from UUID to int
    metric_name: str
    current_value: Decimal
    limit_value: Decimal
    percent_used: Decimal
    status: RiskStatus
    last_updated: datetime
    created_at: datetime

    @field_serializer("current_value", "limit_value", "percent_used")
    def _serialize_decimal_as_number(self, v: Decimal | None) -> float | None:
        return _decimal_to_float(v)


class RiskViolation(BaseModel):
    """Risk limit violation record"""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    user_id: int  # Changed from UUID to int
    metric_name: str
    violation_type: ViolationType
    current_value: Decimal
    limit_value: Decimal
    severity: Severity
    message: str
    resolved: bool
    resolved_at: datetime | None
    created_at: datetime

    @field_serializer("current_value", "limit_value")
    def _serialize_decimal_as_number(self, v: Decimal | None) -> float | None:
        return _decimal_to_float(v)


class RiskLimit(BaseModel):
    """User risk limit configuration"""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    user_id: int  # Changed from UUID to int
    limit_name: str
    limit_value: Decimal
    warning_threshold: Decimal
    critical_threshold: Decimal
    enabled: bool
    created_at: datetime
    updated_at: datetime
    updated_by: int | None  # Changed from UUID to int

    @field_serializer("limit_value", "warning_threshold", "critical_threshold")
    def _serialize_decimal_as_number(self, v: Decimal | None) -> float | None:
        return _decimal_to_float(v)


class EmergencyStop(BaseModel):
    """Emergency stop record"""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    user_id: int  # Changed from UUID to int
    triggered_by: int  # Changed from UUID to int
    reason: str
    strategies_stopped: int
    orders_cancelled: int
    status: EmergencyStopStatus
    triggered_at: datetime
    resolved_at: datetime | None
    resolved_by: int | None  # Changed from UUID to int


class RiskDashboardSummary(BaseModel):
    """Summary statistics for risk dashboard"""
    total_metrics: int
    breached_metrics: int
    warning_metrics: int
    critical_metrics: int
    recent_violations_count: int
    is_emergency_active: bool


class RiskDashboardData(BaseModel):
    """Complete risk dashboard data"""

    model_config = ConfigDict()

    metrics: list[RiskMetric]
    recent_violations: list[RiskViolation]
    active_limits: list[RiskLimit]
    emergency_status: EmergencyStop | None
    summary: RiskDashboardSummary
    last_updated: datetime


# ===========================
# WEBSOCKET EVENT MODELS
# ===========================


class RiskMetricUpdate(BaseModel):
    """WebSocket event for risk metric update"""

    type: str = "risk_metric_update"
    data: RiskMetric


class RiskViolationAlert(BaseModel):
    """WebSocket event for new risk violation"""

    type: str = "risk_violation_alert"
    data: RiskViolation


class EmergencyStopEvent(BaseModel):
    """WebSocket event for emergency stop"""

    # §13.3 FIX: Match frontend type string exactly
    type: str = "emergency_stop_event"
    data: EmergencyStop
